src/main/java/com/ibm/watson/modelmesh/ModelMeshApi.java

```diff
@@ -784,12 +788,25 @@ public void onHalfClose() {
                     call.close(status, emptyMeta());
                     Metrics metrics = delegate.metrics;
                     if (metrics.isEnabled()) {
-                        if (isVModel) {
-                            metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
-                                    status.getCode(), reqSize, respSize, "", Iterables.toString(modelIds));
-                        } else {
-                            metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
-                                    status.getCode(), reqSize, respSize, Iterables.toString(modelIds), "");
+                        Iterator<String> midIt = modelIds.iterator();
+                        while (midIt.hasNext()) {
+                            if (isVModel) {
+                                String mId = null;
+                                String vmId = midIt.next();
+                                try {
+                                    mId = vmm().resolveVModelId(midIt.next(), mId);
+                                    metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, mId, vmId);
+                                }
+                                catch (Exception e) {
+                                    logger.error("Could not resolve model id for vModelId" + vmId, e);
+                                    metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, "", vmId);
+                                }
+                            } else {
+                                metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, midIt.next(), "");
+                            }
```

Comment:
@VedantMahabaleshwarkar thinking more here, it's probably too complicated to log the metrics in the multi-id case. This functionality was for a fairly unique use and I doubt it's actually being used by anyone anymore.

So I would say adjust the logic further up in this method to ensure the modelId / vModelId variables are set by the time it reaches this finally block.

I.e. add a local vModel variable and ensure that this is set in the vModel case. Maybe add a resolvedModelId variable too, which can be set to the modelId above in the modelId case and otherwise set within this finally block above, from the new ThreadLocal mentioned in the other comment (prior to the call to `clearThreadLocals()`). I'm suggesting to call it resolvedModelId rather than modelId since the latter currently refers to ether modelId or vModelId.

Maybe in the multi-id case we leave both these vars as null and then the logic here will know to just not record any metric at all.

Reply:
@VedantMahabaleshwarkar WDYT about this suggestion?

Reply:
@njhill does the change look good to you? I separated out the single-model and multi-model case and the rest of the logic is in line with your suggestion I think